lnwallet/rebroadcaster.go

```diff
@@ -0,0 +1,335 @@
+	// will be used to broadcast the higher fee re-signed transaction to
+	// ensure it's accepted into the backend's mempool.
+	Broadcast func(*wire.MsgTx) error
+}
+
+// TxRebroadcaster is a helper struct that aids us in rebroadcasting
+// transactions with insufficient fees in order to propagate throughout the
+// network.
+type TxRebroadcaster struct {
+	cfg     TxRebroadcasterCfg
+	errChan chan error
+
+	mu          sync.Mutex
+	numAttempts int
+	lastFee     btcutil.Amount
+	lastTx      *wire.MsgTx
+}
+
+// NewTxRebroadcaster creates a new rebroadcaster using the given configuration.
+func NewTxRebroadcaster(cfg TxRebroadcasterCfg) (*TxRebroadcaster, error) {
+	// Validate the constraints specified within the config.
+	if cfg.BumpPercentage <= 0 {
+		return nil, errors.New("bump percentage must be set")
+	}
+	if cfg.MaxAttempts <= 0 && cfg.MaxFee <= 0 {
+		return nil, errors.New("either MaxAttempts or MaxFee must be set")
+	}
+	if cfg.DustLimit <= 0 {
+		return nil, errors.New("dust limit must be set")
+	}
+
+	// Ensure this is a valid transaction.
+	tx := btcutil.NewTx(cfg.Tx)
+	if err := blockchain.CheckTransactionSanity(tx); err != nil {
+		return nil, err
+	}
+
+	return &TxRebroadcaster{
+		cfg:     cfg,
+		errChan: make(chan error, 1),
+	}, nil
+}
+
+// Rebroadcast kicks off the automatic rebroadcast process. The quit channel can
+// be used to signal the rebroadcaster to terminate early.
+func (r *TxRebroadcaster) Rebroadcast(quit <-chan struct{}) <-chan error {
+	go r.rebroadcast(quit)
+
+	return r.errChan
+}
+
+// rebroadcast is the main handler of the rebroadcaster. For every attempt, the
+// current fee is bumped by the specified percentage. Each new transaction is
+// then signed and broadcast to the network. If the fee is not sufficient, the
+// whole process will be done again until either we can read from the quit
+// channel, we've reached the maximum fee allowed, or we've reached the maximum
+// number of attempts allowed.
+func (r *TxRebroadcaster) rebroadcast(quit <-chan struct{}) {
+	// We'll start off by computing the current fee of the transaction.
+	// We do this by subtracting the total value of the outputs from the
+	// total value of the inputs.
+	inputAmt := int64(0)
+	for _, input := range r.cfg.Tx.TxIn {
+		amount, err := r.cfg.FetchInputAmount(&input.PreviousOutPoint)
+		if err != nil {
+			r.errChan <- fmt.Errorf("unable to fetch utxo used as "+
+				"input: %v", err)
+			return
+		}
+
+		inputAmt += int64(amount)
+	}
+
+	outputAmt := int64(0)
+	for _, output := range r.cfg.Tx.TxOut {
+		outputAmt += output.Value
+	}
+
+	currentFee := inputAmt - outputAmt
+	changeOutput := r.cfg.Tx.TxOut[r.cfg.ChangeOutputIndex]
+
+	r.mu.Lock()
+	r.lastFee = btcutil.Amount(currentFee)
+	r.mu.Unlock()
+
+	// With the current fee computed, we can begin our rebroadcast attempts
+	// by bumping up the transaction's fee at every iteration.
+	for {
+		r.mu.Lock()
+		numAttempts := r.numAttempts
+		currentFee := r.lastFee
```

Comment:
Is it necessary to do this in three locations? Before the loop, at the beginning of the loop and at the end of the loop.